### PR TITLE
lib/vector/Vlib: Fix copy into buffer size issue in color_read.c

### PR DESCRIPTION
**Pull request by ShubhamDesai**
This pull request resolves a buffer overflow issue detected by Coverity Scan (CID 1208238).
strcpy is replaced with G_strlcpy
sprintf is replaced with snprintf

### CODE REVIEW

File: lib/vector/Vlib/color_read.c [hunk 1]
```diff
@@ -49,7 +49,9 @@ int Vect_read_colors(const char *name, const char *mapset,
     if (colors)
         Rast_init_colors(colors);
 
-    strcpy(xname, name);
+    if (G_strlcpy(xname, name, sizeof(xname)) >= sizeof(xname)) {
+        G_fatal_error(_("Vector map name <%s> is too long"), name);
```
Comment (nilason) [inline lib/vector/Vlib/color_read.c:53]: As the function do have a return value of failure (and is part of the GRASS library), let us issue a warning instead and return `-1`.

File: lib/vector/Vlib/color_read.c [hunk 2]
```diff
@@ -58,12 +60,12 @@ int Vect_read_colors(const char *name, const char *mapset,
 
     if (strcmp(mapset, G_mapset()) == 0) {
         /* look for the regular color table */
-        sprintf(buf, "%s/%s", GV_DIRECTORY, name);
+        snprintf(buf, sizeof(buf), "%s/%s", GV_DIRECTORY, name);
```
Comment (nilason) [inline lib/vector/Vlib/color_read.c:63]: It is safe to assume there will be no truncation in this case, but make this "exclicit" to the compiler.

```suggestion
        (void)snprintf(buf, sizeof(buf), "%s/%s", GV_DIRECTORY, name);
```

File: lib/vector/Vlib/color_read.c [hunk 3]
```diff
@@ -58,12 +60,12 @@ int Vect_read_colors(const char *name, const char *mapset,
...
         ret = Rast__read_colors(buf, GV_COLR_ELEMENT, mapset, colors);
     }
     else {
         /* look for secondary color table in current mapset */
-        sprintf(buf, "%s/%s", GV_COLR2_DIRECTORY, mapset);
+        snprintf(buf, sizeof(buf), "%s/%s", GV_COLR2_DIRECTORY, mapset);
```
Comment (nilason) [inline lib/vector/Vlib/color_read.c:68]: Same as above.

**Review by nilason — CHANGES_REQUESTED** [review]
Some minor suggestions.

Note, the use of this function are in many cases faulty. The return value is often treated as a bool, but that is incorrect in case of error. The boolean expression of `-1` is true! Only `0` is false, everything else is true.

Addressing that is out-of-scope for this PR, but if you feel like it you may fix those cases in another PR. (https://github.com/search?q=repo%3AOSGeo%2Fgrass%20Vect_read_colors&type=code)

**Comment by ShubhamDesai** [comment]
> Some minor suggestions.
> 
> Note, the use of this function are in many cases faulty. The return value is often treated as a bool, but that is incorrect in case of error. The boolean expression of `-1` is true! Only `0` is false, everything else is true.
> 
> Addressing that is out-of-scope for this PR, but if you feel like it you may fix those cases in another PR. (https://github.com/search?q=repo%3AOSGeo%2Fgrass%20Vect_read_colors&type=code)

So you are suggesting to change -1 with 0?

**Comment by nilason** [comment]
> > Some minor suggestions.
> > Note, the use of this function are in many cases faulty. The return value is often treated as a bool, but that is incorrect in case of error. The boolean expression of `-1` is true! Only `0` is false, everything else is true.
> > Addressing that is out-of-scope for this PR, but if you feel like it you may fix those cases in another PR. (https://github.com/search?q=repo%3AOSGeo%2Fgrass%20Vect_read_colors&type=code)
> 
> So you are suggesting to change -1 with 0?

No, the function `Vect_read_colors()` should not change, but calls using the function must be adapted to the possibility of a return value of `-1` (= error), for example:

https://github.com/OSGeo/grass/blob/b4187339ee5bcf91db6425931d0774dd0f48ad33/misc/m.nviz.image/vector.c#L145-L147

`have_colors` is in this case true also when errored (`-1`).

A possible solution can be found in:

https://github.com/OSGeo/grass/blob/b4187339ee5bcf91db6425931d0774dd0f48ad33/vector/v.colors.out/main.c#L81-L89

(although the line 85, now that `-1` is excluded, could be simplified to `if (!ret) {`)

File: lib/vector/Vlib/color_read.c (hunk 1 above)
Comment (ShubhamDesai) [reply lib/vector/Vlib/color_read.c]: Done

File: lib/vector/Vlib/color_read.c (hunk 2 above)
Comment (ShubhamDesai) [reply lib/vector/Vlib/color_read.c]: Done

File: lib/vector/Vlib/color_read.c (hunk 3 above)
Comment (ShubhamDesai) [reply lib/vector/Vlib/color_read.c]: Done

**Review by nilason — APPROVED** [review]
Thanks!

**Pull request merged**